spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java

```diff
@@ -558,6 +637,63 @@ private void sendReplyForMessageSource(Object result, String topic, Message<?> s
 		this.replyTemplate.send(builder.build());
 	}
 
+	protected void asyncSuccess(@Nullable Object result, String replyTopic, Message<?> source, boolean returnTypeMessage) {
+		if (result == null) {
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug("Async result is null, ignoring");
+			}
+		}
+		else {
+			sendResponse(result, replyTopic, source, returnTypeMessage);
+		}
+	}
+
+	protected void acknowledge(@Nullable Acknowledgment acknowledgment) {
+		if (acknowledgment != null) {
+			acknowledgment.acknowledge();
+		}
+	}
+
+	protected void asyncFailure(Object request, @Nullable Acknowledgment acknowledgment, Consumer<?, ?> consumer,
+			Throwable t, Message<?> source) {
+		try {
+			handleException(request, acknowledgment, consumer, source,
+					new ListenerExecutionFailedException(createMessagingErrorMessage(
+							"Async Fail", source.getPayload()), t));
+			return;
+		}
+		catch (Exception ex) {
+		}
+		this.logger.error(t, () -> "Future, Mono, or suspend function was completed with an exception for " + source);
+		acknowledge(acknowledgment);
+	}
+
+	protected void handleException(Object records, @Nullable Acknowledgment acknowledgment, Consumer<?, ?> consumer,
+			Message<?> message, ListenerExecutionFailedException e) {
+
+		if (this.errorHandler != null) {
```

Comment:
> I mean it feel like this call has to be removed from here

When this.errorHandler != null, need acknowledgment skip error record.

Reply:
OK. The idea is to ack the record we have just handled via `errorHandler`.
Isn't that this error handler responsibility then?
I'm not sure though what we should do otherwise, when ` this.errorHandler == null`...

Reply:
> When this.errorHandler != null, need acknowledgment skip error record.

Sorry, misdescription, logic is when `this.errorHandler == null`, need acknowledgment skip error record.